Log a warning if 'akka.remote.log-frame-size-exceeding' contains a bad value

danbim opened the pull request:
Currently, if 'akka.remote.log-frame-size-exceeding' is accidentally set
to an invalid value (e.g. 'on') the developer does not get warned but
remote communication fails with a not-very-helpful error message:

```
[INFO] [06/29/2018 15:07:41.657] [run-main-0] [akka.remote.Remoting] Starting remoting
[INFO] [06/29/2018 15:07:41.847] [run-main-0] [akka.remote.Remoting] Remoting started; listening on addresses :[akka.tcp://akkaremotenull@localhost:2553]
[INFO] [06/29/2018 15:07:41.849] [run-main-0] [akka.remote.Remoting] Remoting now listens on addresses: [akka.tcp://akkaremotenull@localhost:2553]
[WARN] [06/29/2018 15:07:41.986] [akkaremotenull-akka.remote.default-remote-dispatcher-5] [akka.tcp://akkaremotenull@localhost:2553/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fakkaremotenull%40localhost%3A2552-0] Association with remote system [akka.tcp://akkaremotenull@localhost:2552] has failed, address is now gated for [5000] ms. Reason: [akka://akkaremotenull/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fakkaremotenull%40localhost%3A2552-0/endpointWriter: exception during creation] Caused by: [null]
[INFO] [06/29/2018 15:07:41.993] [akkaremotenull-akka.actor.default-dispatcher-3] [akka://akkaremotenull/deadLetters] Message [io.ino.akkaremotenull.Greeter$Greet] from Actor[akka://akkaremotenull/temp/$a] to Actor[akka://akkaremotenull/deadLetters] was not delivered. [1] dead letters encountered. If this is not an expected behavior, then [Actor[akka://akkaremotenull/deadLetters]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

This commit fixes the issue by a) printing a warning message and b)
ignoring the config setting.

danbim committed:
Log a warning if 'akka.remote.log-frame-size-exceeding' contains a bad value

Currently, if 'akka.remote.log-frame-size-exceeding' is accidentally set
to an invalid value (e.g. 'on') the developer does not get warned but
remote communication fails with a not-very-helpful error message:

```
[INFO] [06/29/2018 15:07:41.657] [run-main-0] [akka.remote.Remoting] Starting remoting
[INFO] [06/29/2018 15:07:41.847] [run-main-0] [akka.remote.Remoting] Remoting started; listening on addresses :[akka.tcp://akkaremotenull@localhost:2553]
[INFO] [06/29/2018 15:07:41.849] [run-main-0] [akka.remote.Remoting] Remoting now listens on addresses: [akka.tcp://akkaremotenull@localhost:2553]
[WARN] [06/29/2018 15:07:41.986] [akkaremotenull-akka.remote.default-remote-dispatcher-5] [akka.tcp://akkaremotenull@localhost:2553/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fakkaremotenull%40localhost%3A2552-0] Association with remote system [akka.tcp://akkaremotenull@localhost:2552] has failed, address is now gated for [5000] ms. Reason: [akka://akkaremotenull/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fakkaremotenull%40localhost%3A2552-0/endpointWriter: exception during creation] Caused by: [null]
[INFO] [06/29/2018 15:07:41.993] [akkaremotenull-akka.actor.default-dispatcher-3] [akka://akkaremotenull/deadLetters] Message [io.ino.akkaremotenull.Greeter$Greet] from Actor[akka://akkaremotenull/temp/$a] to Actor[akka://akkaremotenull/deadLetters] was not delivered. [1] dead letters encountered. If this is not an expected behavior, then [Actor[akka://akkaremotenull/deadLetters]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

This commit fixes the issue by a) printing a warning message and b)
ignoring the config setting.

lightbend-cla-validator commented:
Hi @danbim,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

danbim commented:
I created a little hello world app to showcase the issue: https://github.com/danbim/akka-remote-null

danbim committed:
Merge branch 'master' into wip-warn-on-logFrameSizeExceeding-parse-error

johanandren commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -56,35 +59,48 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
+
+  private val logFrameSizeExceeding: Try[Int] = Try(
+    system.settings.config.getBytes("akka.remote.log-frame-size-exceeding").toInt
+  )
```
A try-catch would be better than a `Try` in this case, this will cause even message to allocate an extra function and go through an extra dispatch to that function.

danbim replied:
Hey there :). Not perfectly sure what you mean. "... will cause even message ..." Do you mean every message? I'd expect this code to be executed only once for the first message, am I mistaken here?
Or are you merely referring to the `logFrameSizeExceeding.failed.foreach` that could be avoided if using a try-catch. In that case I'd suggest to make it a `logFrameSizeExceeding: Option[Int]`.

johanandren commented:
OK TO TEST

patriknw commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -56,35 +59,48 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
+    val value = system.settings.config.getAnyRef("akka.remote.log-frame-size-exceeding")
+    val msg = "Caught exception of type \"" + e.getClass.getSimpleName + "\" " +
+              "while parsing \"akka.remote.log-frame-size-exceeding\". Expected a value in bytes, " +
+              "got: \"" + value + "\". For reference, check https://github.com/lightbend/config/blob/master/HOCON.md"
```
we often use `[]` to surround such values in log and exception messages

```
s"got [$value]. For reference, check https://github.com/lightbend/config/blob/master/HOCON.md"
```

danbim committed:
Adress review comments

Allows value 'off' for 'akka.remote.log-frame-size-exceeding' to be more
tolerant and now logs the original error message of the ConfigException
caught when parsing an invalid value:

```
[WARN] [07/05/2018 09:13:19.544] [akkaremotenull-akka.remote.default-remote-dispatcher-13] [RemoteMetricsOn(akka://akkaremotenull)] application.conf @ file:/Users/danbim/coding/akka-remote-null/target/scala-2.12/classes/application.conf: 19: Invalid value at 'akka.remote.log-frame-size-exceeding': No number in size-in-bytes value 'test'. For reference, check https://github.com/lightbend/config/blob/master/HOCON.md
```

johanandren commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -54,53 +55,59 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
+          log.warning(e.getMessage + ". For reference, check https://github.com/lightbend/config/blob/master/HOCON.md")
+          Int.MaxValue
+      }
+    }
```
This is fine, but just for the record there is a neater way to do this, using `conf.getValue(key)` and then matching on the returned type.

johanandren commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -54,53 +55,59 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
-          case x: RouterEnvelope        ⇒ x.message.getClass
-          case _                        ⇒ msg.getClass
-        }
+    if (payloadBytes >= logFrameSizeExceeding) {
```
if we put a check against Int.MaxValue first here I think the JIT can optimize away the entire block when the setting is off, since that fact will never change, instead of having to check for every element size.

johanandren reviewed (commented):
Lookin' good, the second comment I think should be fixed but the first you can do as you wish with.

danbim committed:
Check logFrameSizeExceeding against Int.MaxValue to allow JIT optimizations

johanandren reviewed (approved):
LGTM, thanks!

patriknw commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -51,18 +55,38 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
   private val log = Logging(system, this.getClass)
+
+  private val logFrameSizeExceeding: Int = {
+    if (system.settings.config.getString(LogFrameSizeExceedingConfigKey) == "off") {
```
this can be removed, it can't happen, because if "off" then `RemoteMetricsOff` is used instead of `RemoteMetricsOn`

(and it's anyway missing a `toLowerCase`)

patriknw commented on akka-remote/src/main/scala/akka/remote/RemoteMetricsExtension.scala:
```diff
@@ -51,18 +55,38 @@ private[akka] class RemoteMetricsOff extends RemoteMetrics {
+private object RemoteMetricsOn {
+  val LogFrameSizeExceedingConfigKey = "akka.remote.log-frame-size-exceeding"
+}
+
```
I think a better way to do this is to read the config in `RemoteSettings`. Make it an `Option[Int]`. Then it will fail fast when the ActorSystem is started. We usually think it's better to fail the startup completely rather than falling back to a silent behavior.

Then in `RemoteMetricsExtension.createExtension` the `RemoteSettings` can be accessed via `RARP(system).provider.remoteSettings`.

danbim committed:
Load 'akka.remote.log-frame-size-exceeding' in RemoteSettings

danbim commented:
@patriknw I addressed your suggestions, totally makes sense and improves user experience. You might want to take another look.

chbatey reviewed (approved):
LGTM

chbatey commented:
thanks @danbim sorry this took a while to merge
